src/SAWScript/Crucible/LLVM/Setup/Value.hs

```diff
@@ -0,0 +1,350 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE PatternGuards #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module SAWScript.Crucible.LLVM.Setup.Value
+  ( LLVM
+    -- * LLVMMethodId
+  , LLVMMethodId(..)
+  , llvmMethodParent
+  , llvmMethodName
+    -- * LLVMAllocSpec
+  , LLVMAllocSpec(..)
+  , LLVMAllocSpecInit(..)
+  , allocSpecType
+  , allocSpecAlign
+  , allocSpecMut
+  , allocSpecMd
+  , allocSpecBytes
+  , allocSpecFresh
+  , allocSpecInit
+    -- * LLVMModule
+  , LLVMModule -- abstract
+  , modFilePath
+  , modAST
+  , modTrans
+  , loadLLVMModule
+  , showLLVMModule
+    -- * CrucibleContext
+  , LLVMCrucibleContext(..)
+  , ccLLVMSimContext
+  , ccLLVMModule
+  , ccLLVMGlobals
+  , ccBasicSS
+  , ccBackend
+    -- * PointsTo
+  , LLVMPointsTo(..)
+  , LLVMPointsToValue(..)
+    -- * AllocGlobal
+  , LLVMAllocGlobal(..)
+  , ppAllocGlobal
+    -- * ResolvedState
+  , LLVMResolvedState(..)
+  , ResolvedPath
+  , ResolvedPathItem(..)
+  , emptyResolvedState
+  , rsAllocs
+  , rsGlobals
+    -- * @LLVMPtr@
+  , LLVMPtr
+  ) where
+
+import           Control.Lens
+import           Data.Map ( Map )
+import qualified Data.Map as Map
+import qualified Data.Text as Text
+import           Data.Type.Equality (TestEquality(..))
+import qualified Prettyprinter as PPL
+import qualified Text.LLVM.AST as L
+import qualified Text.LLVM.PP as L
+import qualified Text.PrettyPrint.HughesPJ as PP
+
+import qualified Data.LLVM.BitCode as LLVM
+
+import           Data.Parameterized.Some (Some(Some))
+
+import           What4.ProgramLoc (ProgramLoc)
+
+import qualified Lang.Crucible.FunctionHandle as Crucible (HandleAllocator)
+import qualified Lang.Crucible.Simulator.ExecutionTree as Crucible (SimContext)
+import qualified Lang.Crucible.Simulator.GlobalState as Crucible (SymGlobalState)
+
+import           SAWScript.Crucible.Common
+import qualified SAWScript.Crucible.Common.Setup.Value as Setup
+
+import qualified SAWScript.Crucible.LLVM.CrucibleLLVM as CL
+
+import           SAWScript.Proof (TheoremNonce)
+
+import           Verifier.SAW.Rewriter (Simpset)
+import           Verifier.SAW.SharedTerm
+import           Verifier.SAW.TypedTerm
+
+--------------------------------------------------------------------------------
+-- ** Language features
+
+data LLVM (arch :: CL.LLVMArch)
+
+type instance Setup.XSetupNull (LLVM _) = ()
+-- 'True' if this is an LLVM packed struct, 'False' otherwise.
+type instance Setup.XSetupStruct (LLVM _) = Bool
+type instance Setup.XSetupArray (LLVM _) = ()
+type instance Setup.XSetupElem (LLVM _) = ()
+type instance Setup.XSetupField (LLVM _) = ()
+type instance Setup.XSetupCast (LLVM _) = L.Type
```

Comment:
These `XSetupStruct` and `XSetupCast` instances are the big payoffs in this patch. They no longer evaluate to `()`, but rather to `Bool`/`L.Type`. Moreover, this information is _only_ used in the LLVM backend, which is why we are able to, e.g., delete so many uses of "`packed`" in the `SetupStruct` cases for the JVM and MIR backends.